ui/index.d.ts

```diff
@@ -234,3 +236,139 @@ export interface IParsedError {
+
+    /**
+     * Opens an input box to ask the user for input.
+     *
+     * @param options Configures the behavior of the input box.
+     * @throws `UserCancelledError` if the user cancels.
+     * @return A promise that resolves to a string the user provided.
+     */
+    showInputBox(options: InputBoxOptions): Promise<string>;
+
+    /**
+     * Show a warning message.
+     *
+     * @param message The message to show.
+     * @param items A set of items that will be rendered as actions in the message.
+     * @throws `UserCancelledError` if the user cancels.
+     * @return A thenable that resolves to the selected item when being dismissed.
+     */
+    showWarningMessage(message: string, ...items: MessageItem[]): Promise<MessageItem>;
+
+    /**
+     * Shows a file open dialog to the user which allows to select a file
+     * for opening-purposes.
+     *
+     * @param options Options that control the dialog.
+     * @throws `UserCancelledError` if the user cancels.
+     * @returns A promise that resolves to the selected resources.
+     */
+    showOpenDialog(options: OpenDialogOptions): Promise<Uri[]>;
+}
+
+/**
+ * Wrapper class of several `vscode.window` methods that handle user input.
+ */
+export declare class AzureUserInput implements IAzureUserInput {
+    /**
+     * @param persistence Used to persist previous selections in the QuickPick.
+     */
+    public constructor(persistence: Memento);
+
+    public showQuickPick<T extends QuickPickItem>(items: T[] | Thenable<T[]>, options: QuickPickOptions): Promise<T>;
+    public showInputBox(options: InputBoxOptions): Promise<string>;
+    public showWarningMessage(message: string, ...items: MessageItem[]): Promise<MessageItem>;
+    public showOpenDialog(options: OpenDialogOptions): Promise<Uri[]>;
+}
+
+/**
+ * Wrapper class of several `vscode.window` methods that handle user input.
+ * This class is meant to be used for testing in non-interactive mode.
+ */
+export declare class TestUserInput implements IAzureUserInput {
+    /**
+     * @param inputs An ordered array of inputs that will be used instead of interactively prompting in VS Code.
+     */
+    public constructor(inputs: (string | undefined)[]);
+
+    public showQuickPick<T extends QuickPickItem>(items: T[] | Thenable<T[]>, options: QuickPickOptions): Promise<T>;
+    public showInputBox(options: InputBoxOptions): Promise<string>;
+    public showWarningMessage(message: string, ...items: MessageItem[]): Promise<MessageItem>;
+    public showOpenDialog(options: OpenDialogOptions): Promise<Uri[]>;
+}
+
+/**
+ * Provides additional options for QuickPickItems used in Azure Extensions
+ */
+export interface IAzureQuickPickItem<T = undefined> extends QuickPickItem {
+    /**
+     * An optional id to uniqely identify this item items across sessions, used in persisting previous selections
+     * If not specified, a hash of the label will be used
+     */
+    id?: string;
```

Comment:
You think we should rename this persistenceId?

Reply:
I chose just `id` to be consistent with how its done for TreeItems. See the comment here: https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L5233

In both TreeItems and Picks, its used to persist some sort of state, but at the end of the day its just an id for that thing. In other words, we could potentially use it for something other than persistence in the future